pkg/wal/wal.go

```diff
@@ -43,18 +128,450 @@ type WAL interface {
+	}
+	log := &Log{path: path, options: *options}
+	if err := os.MkdirAll(path, os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	if err := log.load(); err != nil {
+		return nil, err
+	}
+	log.runFlushTask()
+	return log, nil
+}
+
+func (log *Log) runFlushTask() {
+	go func() {
+		bufferSize := 0
+		for {
+			timer := time.NewTimer(bufferBatchInterval * time.Millisecond)
+			select {
+			case request := <-log.writeChannel:
+				bufferSize += seriesIDLength + timestampLength + len(request.data)
+				err := log.buffer.write(request)
+				if err != nil {
+					log.l.Error().Err(err).Msg("Fail to write to buffer")
+				}
+				if bufferSize > log.options.BufferSize {
+					// Clone buffer,avoiding to block write.
+					buf := log.buffer
+					// Clear buffer to receive Log request.
+					log.newBuffer()
+					if log.closeFlag {
+						return
+					}
+					log.asyncBatchflush(buf, log.flushChannel)
+				}
+			case <-timer.C:
+				buf := log.buffer
+				log.newBuffer()
+				if log.closeFlag {
+					return
+				}
+				log.asyncBatchflush(buf, log.flushChannel)
+			}
+		}
+	}()
+}
+
+func (log *Log) asyncBatchflush(buffer buffer, flushCh chan string) {
+	go func() {
+		// Convert to byte.
+		var err error
+		bytesBuffer := bytes.NewBuffer([]byte{})
+		err = binary.Write(bytesBuffer, binary.LittleEndian, int64(log.buffer.size))
+		if err != nil {
+			log.l.Error().Err(err).Msg("EntryLength fail to convert to byte")
+		}
+		for seriesID, timestamp := range buffer.timestampMap {
+			count := 0
+			var timeBytes []byte
+			for _, timestamp := range timestamp {
+				time := []byte(timestamp.String())
+				count += len(time)
+				timeBytes = append(timeBytes, time...)
+			}
+			count /= timestampLength
+			entryLength := seriesIDLength + count + len(timeBytes) + binaryLength + len(buffer.valueMap[seriesID])
+			err = binary.Write(bytesBuffer, binary.LittleEndian, int64(entryLength))
+			if err != nil {
+				log.l.Error().Err(err).Msg("EntryLength fail to convert to byte")
+			}
+			err = binary.Write(bytesBuffer, binary.LittleEndian, uint64(seriesID))
+			if err != nil {
+				log.l.Error().Err(err).Msg("SeriesID fail to convert to byte")
+			}
+			err = binary.Write(bytesBuffer, binary.LittleEndian, int32(count))
+			if err != nil {
+				log.l.Error().Err(err).Msg("Count fail to convert to byte")
+			}
+			err = binary.Write(bytesBuffer, binary.LittleEndian, timeBytes)
+			if err != nil {
+				log.l.Error().Err(err).Msg("Timestamp fail to convert to byte")
+			}
+			err = binary.Write(bytesBuffer, binary.LittleEndian, int16(len(buffer.valueMap[seriesID])))
+			if err != nil {
+				log.l.Error().Err(err).Msg("Binary Length fail to convert to byte")
+			}
+			err = binary.Write(bytesBuffer, binary.LittleEndian, buffer.valueMap[seriesID])
+			if err != nil {
+				log.l.Error().Err(err).Msg("Value fail to convert to byte")
+			}
+		}
+
+		// Compression and flush.
+		compressionData := snappy.Encode(nil, bytesBuffer.Bytes())
```

Comment:
Instead of compressing the entire segment, it is recommended to separately encode/compress timestamps and values.

The `pkg/encoding/xor` package offers methods for encoding uint64. These can be used to encode/decode the UNIX epoch (nanosecond) of a timestamp.

For quick compression of values, snappy can be utilized.

By following this approach, there is no need for `bytesBuffer` during flushing. Instead, data can be written directly to `workSegment.file`.